src/net/db.rs

```diff
@@ -1073,6 +1263,149 @@ impl PeerDB {
                     WHERE network_id = ?11 AND addrbytes = ?12 AND port = ?13", args)
             .map_err(db_error::SqliteError)?;
 
+        if let Some(old_peer) = old_peer_opt {
+            let slots = Self::find_peer_slots(
+                tx,
+                neighbor.addr.network_id,
+                &neighbor.addr.addrbytes,
+                neighbor.addr.port,
+            )?;
+            if old_peer.public_key.to_bytes_compressed()
+                != neighbor.public_key.to_bytes_compressed()
+            {
+                // this peer has re-keyed, so it might be a new peer altogether.
+                // require it to re-announce its DBs
+                for slot in slots {
+                    debug!("Peer at slot {} changed; dropping its DBs", slot);
+                    PeerDB::drop_stacker_dbs(tx, slot)?;
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Find a peer's slots
+    fn find_peer_slots(
+        conn: &Connection,
+        network_id: u32,
+        addrbytes: &PeerAddress,
+        port: u16,
+    ) -> Result<Vec<u32>, db_error> {
+        let slots = PeerDB::peer_slots(conn, network_id, addrbytes, port)?;
+        let mut ret = vec![];
+        for slot in &slots {
+            let used_slot = PeerDB::has_peer_at(conn, network_id, *slot)?;
```

Comment:
See https://github.com/stacks-network/stacks-blockchain/pull/3552#discussion_r1159257436.  A peer has multiple slots in the DB.  This code is determining whether or not a given slot has a corresponding non-empty row in the DB.

Reply:
I am now kind of more confused by this method.

A peer has multiple slots where it _could_ land -- so `PeerDB::peer_slots` returns 8 (or whatever) results. Then `has_peer_at` in like 1297 just checks if there's _any_ peer at the returned slot.

But in the case of a peer B getting stored in one of peer A's alternates, wouldn't this method return a slot corresponding to a different peer? Why is this method returning a vec of slots at all?

And looking at how this is being used, it seems like this would be undesirable (i.e., Peer A being dropped doesn't mean that Peer B's "stackerdb information" should be deleted, right?)

Reply:
So, why couldn't this method be just the following?

`SELECT slot FROM frontier WHERE network_id = ? AND port = ? AND addrbytes = ?` 

The current method seems to be explicitly trying to include slots that the input peer _isn't_ occupying.

Reply:
Have a look at `try_insert_peer()`.  If _any_ slot generated by `PeerDB::peer_slots()` is _empty_, then the peer can be inserted and that particular slot marked as allocated.  If however there are no slots free, then the caller decides whether or not to evict one of the eight mapped peers from the frontier and replace it with the new peer.

You are indeed correct that this code is buggy.  When I wrote the Stacker DB code, I had forgotten to identify the _single_ slot that represents the peer to update.  Instead, the code erroneously updates all eight slots, even if they don't correspond to the same peer. Whoops!

Reply:
> Have a look at try_insert_peer(). If any slot generated by PeerDB::peer_slots() is empty, then the peer can be inserted and that particular slot marked as allocated. If however there are no slots free, then the caller decides whether or not to evict one of the eight mapped peers from the frontier and replace it with the new peer.

Yeah -- I understand that. My point was that the method `find_peer_slots()` appeared to be buggy, but perhaps also over-engineered. Why can't the `slot` field of `frontier` just be selected, and then `find_peer_slots()` returns a single slot? I understand a peer could be inserted in one of multiple slots, but it is never _actually resident_ in more than one, and from my understanding the intent of this method is to get the slot that a peer is actually resident in. Can you add a test coverage for the would-have-been buggy case here? (i.e., a test that would fail with the current implementation, but passes once this method is fixed?)

Reply:
Yes, you're correct.  The code now returns a single slot and has been renamed to `find_peer_slot()`.